src/python-samplerate.cpp

```diff
@@ -0,0 +1,570 @@
+    if (_state == nullptr) error_handler(_err_num);
+  }
+
+  void _destroy() {
+    if (_state != nullptr) {
+      src_delete(_state);
+      _state = nullptr;
+    }
+  }
+
+ public:
+  CallbackResampler(const callback_t &callback_func, double ratio,
+                    const py::object &converter_type, size_t channels)
+      : _callback(callback_func),
+        _ratio(ratio),
+        _converter_type(get_converter_type(converter_type)),
+        _channels(channels) {
+    _create();
+  }
+
+  // copy constructor
+  CallbackResampler(const CallbackResampler &r)
+      : _callback(r._callback),
+        _ratio(r._ratio),
+        _converter_type(r._converter_type),
+        _channels(r._channels) {
+    _state = src_clone(r._state, &_err_num);
+    if (_state == nullptr) error_handler(_err_num);
+  }
+
+  ~CallbackResampler() { _destroy(); }
+
+  void set_buffer(const np_array_f32 &new_buf) { _current_buffer = new_buf; }
+  size_t get_channels() { return _channels; }
+
+  np_array_f32 callback(void) {
+    auto input = _callback();
+
+    auto inbuf = input.request();
+    if (_buffer_ndim == 0) _buffer_ndim = inbuf.ndim;
+
+    _current_buffer = input;
+    return input;
+  }
+
+  py::array_t<float, py::array::c_style> read(size_t frames) {
+    // allocate output array
+    std::vector<size_t> out_shape{frames, _channels};
+    auto output = py::array_t<float, py::array::c_style>(out_shape);
+    py::buffer_info outbuf = output.request();
+
+    if (_state == nullptr) _create();
+
+    // read from the callback
+    size_t output_frames_gen = src_callback_read(
+        _state, _ratio, (long)frames, static_cast<float *>(outbuf.ptr));
+
+    // check error status
+    if (output_frames_gen == 0) {
+      _err_num = src_error(_state);
+      error_handler(_err_num);
+    }
+
+    // if there is only one channel and the input array had only on dimension
+    // we also output a 1D array
+    if (_channels == 1 && _buffer_ndim == 1) {
+      out_shape.pop_back();
+      output = py::array_t<float, py::array::c_style>(
+          out_shape, static_cast<float *>(outbuf.ptr));
+    }
+
+    // create a shorter view of the array
+    if (output_frames_gen < frames) {
+      out_shape[0] = output_frames_gen;
+      output = py::array_t<float, py::array::c_style>(
+          out_shape, static_cast<float *>(outbuf.ptr));
+    }
+
+    return output;
+  }
+
+  void set_starting_ratio(double new_ratio) {
+    _err_num = src_set_ratio(_state, new_ratio);
+    error_handler(_err_num);
+    _ratio = new_ratio;
+  }
+
+  void reset() {
+    _err_num = src_reset(_state);
+    error_handler(_err_num);
+  }
+
+  CallbackResampler clone() const { return CallbackResampler(*this); }
+  CallbackResampler &__enter__() { return *this; }
+  void __exit__(const py::object &exc_type, const py::object &exc,
+                const py::object &exc_tb) {
+    _destroy();
+  }
+};
+
+long the_callback_func(void *cb_data, float **data) {
```

Comment:
I'd move this into an anonymous namespace to guard against ODR violations.

Reply:
It is now part of the `samplerate` namespace. How would that work to put it in an anonymous namespace ?

I have tried to
```
namespace {
long the_callback_func(void *cb_data, float **data);
}
```
and it compiles, but I get the following warning:
```
src/python-samplerate.cpp:64:6: warning: function 'samplerate::(anonymous namespace)::the_callback_func' has internal linkage but is not defined [-Wundefined-internal]
long the_callback_func(void *cb_data, float **data);
     ^
/Users/robinscheibler/Documents/software/python-samplerate/src/python-samplerate.cpp:220:26: note: used here
        src_callback_new(the_callback_func, _converter_type, (int)_channels,
                         ^
```
Is that normal ? Is there another to do it ?

Reply:
Actually, there is also a symbol not found error at runtime, so I don't think this is ok.

Reply:
You must put both declaration and definition into the anonymous namespace, then it should work. I don't think anything here speaks against just definition and declaration in one go right above the definition of the `CallbackResampler` class.

Reply:
Thanks. Actually, it had to go after the declaration of `CallbackResampler` because the callback func references the class itself. So I had to take out the declaration of the `_create` method and put it after the callback function, since it requires it. I put a comment in the code explaining.
You can check it in the code.

Reply:
`_create` only needed a forward declaration of the callback. I moved it back into the class definition for consistency.